feat(systemctl): 实现服务管理器初始化和命令执行

gcsong023 opened the pull request:
#### What this PR does / why we need it?
实现对 systemd、openrc 和 sysvinit 三种服务管理器的支持，提高系统运行的兼容性
#### Summary of your change
- 新增 systemctl 包，实现对 systemd、openrc 和 sysvinit 三种服务管理器的支持
- 添加服务状态检查、启动、停止、重启和启用/禁用功能
- 实现服务发现和智能服务名处理
- 添加配置文件查看功能
- 优化错误处理和日志记录
- 调整其他相关文件对接口的调用
#### Please indicate you've done the following:

- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Made sure commit message follow the rule of [Conventional Commits specification](https://www.conventionalcommits.org/).
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed.

gcsong023 committed:
feat(systemctl): 实现服务管理器初始化和命令执行

- 新增 systemctl 包，实现对 systemd、openrc 和 sysvinit 三种服务管理器的支持
- 添加服务状态检查、启动、停止、重启和启用/禁用功能
- 实现服务发现和智能服务名处理
- 添加配置文件查看功能
- 优化错误处理和日志记录

gcsong023 committed:
refactor(system): 重构系统服务管理逻辑

- 引入 systemctl 工具包以统一处理系统服务
- 优化服务状态获取、配置文件路径解析等逻辑
- 重构 HostToolService 中的 GetToolStatus 方法
- 更新 DockerService、SettingService 等相关服务的处理方式
- 调整快照创建和恢复过程中的服务处理逻辑

gcsong023 committed:
feat(utils): 添加目录复制功能并优化文件复制逻辑

- 新增 CopyDirs 函数，用于复制整个目录及其内容
- 添加对符号链接的复制支持
- 实现通用的 Copy 函数，根据文件类型自动选择 CopyFile 或 CopyDirs
- 在 CopyFile 函数中增加对源文件是目录的检查和错误提示

gcsong023 committed:
refactortoolbox: 重构 Fail2ban 和 Pure-FTPd 的管理逻辑

- 优化了 Fail2ban 和 Pure-FTPd 的启动、停止、重启等操作的实现
- 改进了 Fail2ban 版本信息的获取方法
- 统一了错误处理和日志记录的格式
- 调整了部分导入的包，提高了代码的可维护性

gcsong023 committed:
build: 禁用 CGO 以提高构建性能和兼容性

- 在 Linux 后端构建命令中添加 CGO_ENABLED=0 环境变量
- 此修改可以提高构建速度，并确保生成的二进制文件在没有 C 库依赖的环境中也能运行

gcsong023 committed:
Merge branch '1Panel-dev:dev' into dev

ssongliu commented on backend/utils/systemctl/handle.go:
```diff
@@ -0,0 +1,409 @@
+}
+
+func (h *ServiceHandler) executeAction(action, successMsg string) (ServiceResult, error) {
+	manager := GetGlobalManager()
```
这个方法是不是不用返回 ServiceResult，我看基本也没用到

有错误的话，直接返回 fmt.Errorf("%s operation failed, std: %s, err: %v", string(output), err)

gcsong023 replied:
handle.go文件里面，serviceResult在涉及操作如start、stop、restart等操作时都有用到的，只不过里面的success、Message、Output不是所有场景都有全部使用。比如说RestartService ，实际上调用的是， h.ExecuteAction("restart")，而ExecuteAction是统一构建successMessage后统一调用excuteAcion……

ssongliu commented on backend/app/service/docker.go:
```diff
@@ -451,31 +450,23 @@ func validateDockerConfig() error {
-	return "systemctl", nil
-}
+// func getDockerRestartCommand() (string, error) {
+// 	stdout, err := cmd.Exec("which docker")
```
这个得留着吧，有些用户是使用 snap 安装的 docker，当前方式是否没法满足这种情况

https://github.com/1Panel-dev/1Panel/issues/3242

gcsong023 replied:
是的，当前方式无法满足snap 安装的docker。新提交[b41eb5d](https://github.com/1Panel-dev/1Panel/pull/8380/commits/b41eb5dbe86f5a06385c6ede77ffee2908d6d5dd)中，补充了sanp判定

ssongliu commented on backend/app/service/snapshot_create.go:
```diff
@@ -43,16 +44,34 @@ func snapPanel(snap snapHelper, targetDir string) {
-	if err := common.CopyFile("/usr/local/bin/1panel", path.Join(targetDir, "1panel")); err != nil {
-		status = err.Error()
+	h, err := systemctl.DefaultHandler("1panel")
+
```
/usr/local/bin/1panel
/usr/local/bin/1pctl
/etc/systemd/system/1panel.service
/usr/local/bin/lang
这些文件是安装 1panel 时直接写死的路径，是不是不用再去重新获取

gcsong023 replied:
1panel.service文件是systemd管理器中的名称，安装脚本在sysvinit里面有创建的服务文件名称是1paneld。整体思路是避免硬编码路径，尤其是服务文件的路径。

gcsong023 replied:
不同系统下，1panel.service的路径和名称都是不一样的。另外也是为了避免不同地方的多次硬编码路径。

ssongliu replied:
可以的，刚看到 install 那边已经做了修改

只是在项目中 1Panel/cmd/server/cmd/restore.go 版本回滚的时候，也用到了 1panel.service，方便一起修改一下吗？

gcsong023 replied:
这里：[09c68b7](https://github.com/1Panel-dev/1Panel/pull/8380/commits/09c68b71cb1bc52af4fffd4ea4e2c7a0fcf3909d) 另外，update.go [74cd7e5](https://github.com/1Panel-dev/1Panel/pull/8380/commits/74cd7e58efe0a392d5db29b258fa306a1c6c6073) 进行了修复。

ssongliu commented on backend/app/service/snapshot_recover.go:
```diff
@@ -147,7 +178,9 @@ func (u *SnapshotService) HandleSnapshotRecover(snap model.Snapshot, isRecover b
 	}
-	_, _ = cmd.Exec("systemctl daemon-reload && systemctl restart 1panel.service")
+	if err := systemctl.Restart("1panel"); err != nil {
+		global.LOG.Errorf("restart 1panel service failed: %v", err)
```
这里快照恢复可能会导致 service 文件变化，这里是不是得加一个参数来判断重启之前是否需要 systemctl daemon-reload 来重新加载

gcsong023 replied:
#[4299d54](https://github.com/1Panel-dev/1Panel/pull/8380/commits/4299d54af720e62ed07faae1d81db52a6ff4e819)
增加对systemd管理器判定后，执行systemctl daemon-reload

gcsong023 committed:
refactor(docker): 重构 Docker 服务的重启和操作逻辑

- 添加 isDockerSnapInstalled 函数来判断 Docker 是否通过 Snap 安装
- 在 OperateDocker 和 restartDocker 函数中增加对 Snap 安装的处理
- 移除未使用的 getDockerRestartCommand 函数

gcsong023 committed:
Merge branch '1Panel-dev:dev' into dev

gcsong023 committed:
fix(service): 优化快照恢复后的服务重启逻辑

- 在使用 systemd 管理服务时，增加 daemon-reload 操作以确保服务配置更新
- 重启 1panel 服务，以应用快照恢复的更改

gcsong023 committed:
refactor(server): 支持非 systemd 系统的恢复操作

- 增加 isSystemd 函数判断系统是否为 systemd 类型
- 根据系统类型选择性地恢复服务文件
- 兼容 systemd 和非 systemd 系统的恢复流程

gcsong023 committed:
fix(upgrade): 优化升级过程中的服务重启逻辑

- 移动服务重启逻辑到版本号更新之后，修复因提前重启导致的版本号未更新BUG。
- 在 systemctl 重启之前添加 daemon-reload 命令

gcsong023 committed:
Merge branch '1Panel-dev:dev' into dev
